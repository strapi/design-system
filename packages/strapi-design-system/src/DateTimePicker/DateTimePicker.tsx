/**
 * TODO: refactor this to only use the Inputs and not the whole DatePicker and TimePicker
 * this way the underlying Field components aren't complaining and the inputs _should_
 * not require error and hint being passed.
 */
import { ReactNode } from 'react';

import { DatePicker, DatePickerProps } from '../DatePicker';
import { Field, FieldHint, FieldLabel, FieldError } from '../Field';
import { Flex } from '../Flex';
import { useControllableState } from '../hooks/useControllableState';
import { useId } from '../hooks/useId';
import { TimePicker, TimePickerProps } from '../TimePicker';

export interface DateTimePickerProps
  extends Omit<DatePickerProps, 'value' | 'step' | 'onChange'>,
    Pick<TimePickerProps, 'step'> {
  /**
   * Aria Label used by the DatePicker and the TimePicker inside the DateTimePicker component
   */
  ariaLabel?: string;
  /**
   * Label used to describe the DateTimePicker component
   */
  label: string;
  /**
   * Label Action
   */
  labelAction?: ReactNode;
  onChange?: (date: Date | undefined) => void;
  /**
   * Text used to define the title of the button to open the select options (inside the TimePicker)
   * @default 'select'
   */
  selectButtonTitle?: string;
  /**
   * Value. The Date passed as value
   */
  value?: Date;
}

export const DateTimePicker = ({
  ariaLabel,
  clearLabel = 'clear',
  disabled = false,
  error,
  hint,
  id,
  label,
  labelAction,
  onChange,
  onClear,
  name,
  required = false,
  selectButtonTitle = 'select',
  size = 'M',
  step = 1,
  value,
  ...props
}: DateTimePickerProps) => {
  const initialDate = parseDate(value);
  const [dateValue, setDateValue] = useControllableState<Date | undefined>({
    prop: initialDate,
    onChange,
  });
  const timeValue = dateValue
    ? `${dateValue.getHours()}:${dateValue.getMinutes()}:${dateValue.getSeconds()}`
    : undefined;

  const handleDateChange = (date: Date) => {
    if (timeValue) {
      const [hours, minutes] = timeValue.split(':');
      date.setHours(parseInt(hours, 10));
      date.setMinutes(parseInt(minutes, 10));
    }

    setDateValue(date);
  };

  const handleTimeChange = (time: string) => {
    const dateToSet = dateValue ? new Date(dateValue) : new Date();
    const [hours, minutes] = time.split(':');
    dateToSet.setHours(parseInt(hours, 10));
    dateToSet.setMinutes(parseInt(minutes, 10));

    setDateValue(dateToSet);
  };

  const handleDateClear = () => {
    setDateValue(undefined);

    if (onClear) {
      onClear();
    }
  };

  const handleTimeClear = () => {
    /**
     * Don't ask, i guess its a TS overload thing...
     */
    let dateToSet = dateValue ? new Date(dateValue) : new Date();
    dateToSet.setHours(0);
    dateToSet.setMinutes(0);

    setDateValue(dateToSet);
  };

  const generatedId = useId(id);

  return (
    <Field
      name={name}
      role="group"
      id={generatedId}
      aria-labelledby={generatedId}
      hint={hint}
      error={error}
      required={required}
    >
      <Flex direction="column" alignItems="stretch" gap={1}>
        <FieldLabel action={labelAction}>{label}</FieldLabel>
        <Flex gap={2}>
          {/* @ts-expect-error label should be required here and we should refactor DatePicker */}
          <DatePicker
            ariaLabel={label || ariaLabel}
            selectedDate={dateValue}
            selectedDateLabel={(formattedDate) => `Date picker, current is ${formattedDate}`}
            onChange={handleDateChange}
            error={typeof error === 'string'}
            hint={typeof hint === 'string'}
            required={required}
            size={size}
            onClear={handleDateClear}
            clearLabel={`${clearLabel} date`}
            disabled={disabled}
            id={generatedId}
            {...props}
          />
          {/* @ts-expect-error label should be required here and we should refactor TimePicker */}
          <TimePicker
            size={size}
            aria-label={label || ariaLabel}
            error={typeof error === 'string'}
            hint={typeof hint === 'string'}
            value={timeValue}
            onChange={handleTimeChange}
            /**
             * In reality the time picker is only cleared when there is no date otherwise it's just 00:00
             * and we should show that for accurate UX and therefore you cannot clear it when it's 00:00.
             */
            onClear={timeValue !== undefined && timeValue !== '0:0:0' ? handleTimeClear : undefined}
            clearLabel={`${clearLabel} time`}
            required={required}
            selectButtonTitle={selectButtonTitle}
            disabled={disabled}
            step={step}
            id={generatedId}
          />
        </Flex>
        <FieldHint />
        <FieldError />
      </Flex>
    </Field>
  );
};

const parseDate = (date) => {
  const timestamp = Date.parse(date);

  if (!Number.isNaN(timestamp)) {
    return new Date(timestamp);
  }

  return undefined;
};
