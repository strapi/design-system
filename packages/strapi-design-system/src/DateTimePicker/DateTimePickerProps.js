import React from 'react';
import PropTypes from 'prop-types';

export const DateTimePickerProps = (props) => <div {...props} />;

export const dateTimePickerDefaultProps = {
  ariaLabel: undefined,
  clearLabel: 'clear',
  disabled: false,
  error: undefined,
  hint: undefined,
  label: undefined,
  labelAction: undefined,
  onClear: undefined,
  required: false,
  selectButtonTitle: 'select',
  size: 'M',
  step: 1,
  value: undefined,
};
export const dateTimePickerPropTypes = {
  /**
   * Aria Label used by the DatePicker and the TimePicker inside the DateTimePicker component
   */
  ariaLabel: PropTypes.string,
  /**
   * Clear label used to define the `aria-label` of the clear buttons.
   */
  clearLabel: PropTypes.string,
  /**
   * Disabled attribute
   */
  disabled: PropTypes.bool,
  /**
   * Error is the message shown under the DateTimePicker when something goes wrong.
   */
  error: PropTypes.oneOfType([PropTypes.string, PropTypes.bool]),
  /**
   * This text is used to describe the DateTimePicker component.
   */
  hint: PropTypes.string,
  /**
   * Label used to describe the DateTimePicker component
   */
  label: PropTypes.string,
  /**
   * Label Action
   */
  labelAction: PropTypes.element,
  /**
   * Name, required
   */
  name: PropTypes.string.isRequired,
  /**
   * onChange function, passed from a parent component, it takes the actual date value and it is used inside the different handlers related to the change event for the DatePicker and the TimePicker and also the clear event for the TimePicker
   */
  onChange: PropTypes.func.isRequired,
  /**
   * onClear function
   */
  onClear: PropTypes.func,
  /**
   * Required field boolean
   */
  required: PropTypes.bool,
  /**
   * Text used to define the title of the button to open the select options (inside the TimePicker)
   */
  selectButtonTitle: PropTypes.string,
  /**
   * Size of the Field (possible values S and M)
   */
  size: PropTypes.oneOf(['S', 'M']),
  /**
   * Step value used inside the TimePicker to show the different options for it
   */
  step: PropTypes.number,
  /**
   * Value. The Date passed as value
   */
  value: PropTypes.instanceOf(Date),
};

DateTimePickerProps.defaultProps = dateTimePickerDefaultProps;
DateTimePickerProps.propTypes = dateTimePickerPropTypes;
