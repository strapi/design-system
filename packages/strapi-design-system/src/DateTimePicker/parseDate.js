// Return a Date if the passed param is a date valid format

export const parseDate = (date) => {
  const timestamp = Date.parse(date);

  if (!Number.isNaN(timestamp)) {
    return new Date(timestamp);
  }

  return null;
};
