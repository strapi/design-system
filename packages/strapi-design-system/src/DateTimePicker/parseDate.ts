// Return a Date if the passed param is a date valid format

export const parseDate = (date: string): Date | undefined => {
  const timestamp = Date.parse(date);

  if (!Number.isNaN(timestamp)) {
    return new Date(timestamp);
  }

  /**
   * This should be undefined because our
   * components do not expect to deal with null values.
   */
  return undefined;
};
