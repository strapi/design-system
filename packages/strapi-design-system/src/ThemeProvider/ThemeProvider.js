import * as React from 'react';
import PropTypes from 'prop-types';
import { ThemeProvider as StyledThemeProvider, createGlobalStyle } from 'styled-components';

const GlobalStyle = createGlobalStyle`
 /* http://meyerweb.com/eric/tools/css/reset/ 
   v2.0 | 20110126
   License: none (public domain)
*/
  html,
  body,
  div,
  span,
  applet,
  object,
  iframe,
  h1,
  h2,
  h3,
  h4,
  h5,
  h6,
  p,
  blockquote,
  pre,
  a,
  abbr,
  acronym,
  address,
  big,
  cite,
  code,
  del,
  dfn,
  em,
  img,
  ins,
  kbd,
  q,
  s,
  samp,
  small,
  strike,
  strong,
  sub,
  sup,
  tt,
  var,
  b,
  u,
  i,
  center,
  dl,
  dt,
  dd,
  ol,
  ul,
  li,
  fieldset,
  form,
  label,
  legend,
  table,
  caption,
  tbody,
  tfoot,
  thead,
  tr,
  th,
  td,
  article,
  aside,
  canvas,
  details,
  embed,
  figure,
  figcaption,
  footer,
  header,
  hgroup,
  menu,
  nav,
  output,
  ruby,
  section,
  summary,
  time,
  mark,
  audio,
  video {
    margin: 0;
    padding: 0;
    border: 0;
    font-size: 100%;
    font: inherit;
    vertical-align: baseline;
  }
  /* HTML5 display-role reset for older browsers */
  article,
  aside,
  details,
  figcaption,
  figure,
  footer,
  header,
  hgroup,
  menu,
  nav,
  section {
    display: block;
  }
  body {
    line-height: 1;
  }
  ol,
  ul {
    list-style: none;
  }
  blockquote,
  q {
    quotes: none;
  }
  blockquote:before,
  blockquote:after,
  q:before,
  q:after {
    content: "";
    content: none;
  }
  table {
    border-collapse: collapse;
    border-spacing: 0;
  }
  /* My styles */
  *,
  *:before,
  *:after {
    box-sizing: border-box;
  }

  *:focus-visible {
    box-shadow: 0px 0px 6px ${(props) => props.theme.colors.primary500};
    // Windows High Contrast won't see the bow shadow but transparent will
    // be replaced by a color
    // See https://sarahmhigley.com/writing/whcm-quick-tips/ for more details
    outline: 4px solid transparent;
  }
  /* Focusing the button with a mouse, touch, or stylus will show a subtle drop shadow. */
  *:focus:not(:focus-visible) {
    outline: none;
  }

  body,html{
    height: 100%;
  }
  body {
    font-family: --apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Oxygen, Ubuntu, Cantarell, 'Open Sans', 'Helvetica Neue', sans-serif;
  }
`;

export const ThemeProvider = ({ children, theme }) => {
  return (
    <StyledThemeProvider theme={theme}>
      <GlobalStyle />
      {children}
    </StyledThemeProvider>
  );
};

ThemeProvider.propTypes = {
  children: PropTypes.node.isRequired,
  theme: PropTypes.object.isRequired,
};
