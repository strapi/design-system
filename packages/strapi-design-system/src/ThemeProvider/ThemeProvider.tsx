/**
 * TODO: remove this in v2 in favour of DesignSystemProvider, but this is rather convienient for now.
 */
import * as React from 'react';

import { ThemeProvider as StyledThemeProvider, createGlobalStyle, DefaultTheme } from 'styled-components';

import { LiveRegions } from '../LiveRegions/LiveRegions';
import { lightTheme } from '../themes';

const GlobalStyle = createGlobalStyle`
 /* http://meyerweb.com/eric/tools/css/reset/ 
   v2.0 | 20110126
   License: none (public domain)
*/
  html,
  body,
  div,
  span,
  applet,
  object,
  iframe,
  h1,
  h2,
  h3,
  h4,
  h5,
  h6,
  p,
  blockquote,
  pre,
  a,
  abbr,
  acronym,
  address,
  big,
  cite,
  code,
  del,
  dfn,
  em,
  img,
  ins,
  kbd,
  q,
  s,
  samp,
  small,
  strike,
  strong,
  sub,
  sup,
  tt,
  var,
  b,
  u,
  i,
  center,
  dl,
  dt,
  dd,
  ol,
  ul,
  li,
  fieldset,
  form,
  label,
  legend,
  table,
  caption,
  tbody,
  tfoot,
  thead,
  tr,
  th,
  td,
  article,
  aside,
  canvas,
  details,
  embed,
  figure,
  figcaption,
  footer,
  header,
  hgroup,
  menu,
  nav,
  output,
  ruby,
  section,
  summary,
  time,
  mark,
  audio,
  video {
    margin: 0;
    padding: 0;
    border: 0;
    font-size: 100%;
    font: inherit;
    vertical-align: baseline;
  }
  /* HTML5 display-role reset for older browsers */
  article,
  aside,
  details,
  figcaption,
  figure,
  footer,
  header,
  hgroup,
  menu,
  nav,
  section {
    display: block;
  }
  body {
    line-height: 1;
  }
  ol,
  ul {
    list-style: none;
  }
  blockquote,
  q {
    quotes: none;
  }
  blockquote:before,
  blockquote:after,
  q:before,
  q:after {
    content: "";
    content: none;
  }
  table {
    border-collapse: collapse;
    border-spacing: 0;
  }
  /* My styles */
  *,
  *:before,
  *:after {
    box-sizing: border-box;
  }

  *:focus-visible {
    outline: 2px solid ${({ theme }) => theme.colors.primary600};
    outline-offset: 2px;
  }

  /* Focusing the button with a mouse, touch, or stylus will show a subtle drop shadow. */
  *:focus:not(:focus-visible) {
    outline: none;
  }

  body,html{
    height: 100%;
  }
  body {
    font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Oxygen, Ubuntu, Cantarell, 'Open Sans', 'Helvetica Neue', sans-serif;
  }

  button {
    margin: 0;
    padding: 0;
    border: none;
    background: none;
    color: inherit;
    font-size: 100%;
    cursor: pointer;
    font: inherit;
  }

  textarea,input {
    font: inherit;
  }

  .lock-body-scroll {
    height: 100vh;
    overflow-y: hidden;
  }
`;

export interface ThemeProviderProps {
  children: React.ReactNode;
  theme?: DefaultTheme;
}

export const ThemeProvider = ({ children, theme = lightTheme }: ThemeProviderProps) => {
  return (
    <StyledThemeProvider theme={theme}>
      {children}
      <LiveRegions />
      <GlobalStyle />
    </StyledThemeProvider>
  );
};
