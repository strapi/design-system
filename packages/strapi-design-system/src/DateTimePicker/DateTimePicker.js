import React, { useEffect, useState } from 'react';

import { dateTimePickerDefaultProps, dateTimePickerPropTypes } from './DateTimePickerProps';
import { parseDate } from './parseDate';
import { DatePicker } from '../DatePicker';
import { Field, FieldHint, FieldLabel, FieldError } from '../Field';
import { useId } from '../hooks/useId';
import { Stack } from '../Stack';
import { TimePicker } from '../TimePicker';

export const DateTimePicker = ({
  ariaLabel,
  clearLabel,
  disabled,
  error,
  hint,
  label,
  labelAction,
  onChange,
  onClear,
  name,
  required,
  selectButtonTitle,
  size,
  step,
  value,
  ...props
}) => {
  const initialDate = parseDate(value);
  const [dateValue, setDateValue] = useState(initialDate);
  const [timeValue, setTimeValue] = useState(
    initialDate ? `${initialDate.getHours()}:${initialDate.getMinutes()}:${initialDate.getSeconds()}` : null,
  );

  const handleDateChange = (e) => {
    setDateValue(e);

    let dateToSet;

    if (timeValue) {
      dateToSet = new Date(e);
      dateToSet.setHours(timeValue.split(':')[0]);
      dateToSet.setMinutes(timeValue.split(':')[1]);
    } else {
      dateToSet = new Date(e);
      setTimeValue(`${dateToSet.getHours()}:${dateToSet.getMinutes()}:${dateToSet.getSeconds()}`);
    }
    if (onChange) {
      onChange(dateToSet);
    }
  };

  const handleTimeChange = (e) => {
    setTimeValue(e);

    const dateToSet = dateValue ? new Date(dateValue) : new Date();
    dateToSet.setHours(e.split(':')[0]);
    dateToSet.setMinutes(e.split(':')[1]);

    if (!dateValue) {
      setDateValue(dateToSet);
    }

    if (onChange) {
      onChange(dateToSet);
    }
  };

  const handleDateClear = () => {
    setDateValue(undefined);
    setTimeValue(undefined);
    onClear();
  };

  const handleTimeClear = () => {
    setTimeValue(undefined);

    let dateToSet;

    if (dateValue) {
      dateToSet = new Date(dateValue);
      dateToSet.setHours('00');
      dateToSet.setMinutes('00');
    }

    if (onChange) {
      onChange(dateToSet);
    }
  };

  useEffect(() => {
    if (value && +value !== +dateValue) {
      const parsedData = parseDate(value);
      setDateValue(parsedData);
      setTimeValue(
        parsedData ? `${parsedData.getHours()}:${parsedData.getMinutes()}:${parsedData.getSeconds()}` : null,
      );
    } else if (!value) {
      setDateValue(undefined);
      setTimeValue(undefined);
    }
  }, [value, dateValue]);

  const generatedId = useId(props?.id);

  return (
    <Field
      name={name}
      role="group"
      id={generatedId}
      aria-labelledby={generatedId}
      hint={hint}
      error={error}
      required={required}
    >
      <Stack spacing={1}>
        {label && <FieldLabel action={labelAction}>{label}</FieldLabel>}
        <Stack horizontal spacing={2}>
          <DatePicker
            data-testid="datetimepicker-date"
            // DateTimePicker sharing its Field id to DatePicker component
            id={generatedId}
            name={name}
            ariaLabel={label || ariaLabel}
            error={typeof error === 'string'}
            hint={typeof hint === 'string'}
            selectedDate={dateValue}
            selectedDateLabel={(formattedDate) => `Date picker, current is ${formattedDate}`}
            onChange={handleDateChange}
            required={required}
            size={size}
            onClear={onClear && handleDateClear}
            clearLabel={clearLabel}
            disabled={disabled}
            {...props}
          />
          <TimePicker
            data-testid="datetimepicker-time"
            size={size}
            aria-label={label || ariaLabel}
            error={typeof error === 'string'}
            hint={typeof hint === 'string'}
            value={timeValue}
            onChange={handleTimeChange}
            onClear={onClear && handleTimeClear}
            clearLabel={clearLabel}
            required={required}
            selectButtonTitle={selectButtonTitle}
            disabled={disabled}
            step={step}
          />
        </Stack>
        <FieldHint />
        <FieldError />
      </Stack>
    </Field>
  );
};

DateTimePicker.defaultProps = dateTimePickerDefaultProps;

DateTimePicker.propTypes = dateTimePickerPropTypes;
