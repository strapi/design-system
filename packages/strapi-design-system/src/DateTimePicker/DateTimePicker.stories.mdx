<!--- DateTimePicker.stories.mdx --->

import { useState } from 'react';
import { Meta, Story, Canvas } from '@storybook/addon-docs';
import { ArgsTable } from '@storybook/addon-docs';
import { DateTimePicker } from './DateTimePicker';

<Meta
  title="Design System/Components/DateTimePicker"
  component={DateTimePicker}
  argTypes={{
    label: {
      control: {
        type: 'text',
      },
    },
    value: {
      control: {
        type: 'date',
      },
    },
  }}
/>

# DateTimePicker

Date Time pickers help users selecting a specific date from a calendar and a specific timeslot from a suggested timetable.

**Best practices**

- When you select a date automatically is selected also a time and viceversa (time and date)
- When you clear a date automatically is deleted also the time and viceversa

[View source](https://github.com/strapi/design-system/tree/main/packages/strapi-design-system/src/DateTimePicker)

## Imports

```js
import { DateTimePicker } from '@strapi/design-system/DateTimePicker';
```

## Usage

- A Date Time picker is used to select an unique date. It is not possible to select a date range. It is not possible to write a date from the keyboard.
- When the Date picker is activated, it selects today's date. However, a Date picker can have a null value via the Clear feature.
- Use the Time picker component whenever a specific time needs to be scheduled or indicated. It is only possible to select one single timeslot.

<Canvas>
  <Story name="base">
    {() => {
      const [value, setValue] = useState(undefined);
      return (
        <DateTimePicker
          label={'Date time picker'}
          disabled={false}
          name="datetimepicker"
          onChange={(e) => setValue(e)}
          clearLabel="Clear"
          onClear={() => setValue(undefined)}
          value={value}
          hint="This is a super description"
          selectedDateLabel={() => `Date picker, current is undefined`}
        />
      );
    }}
  </Story>
</Canvas>

## Error

<Canvas>
  <Story name="error">
    {() => {
      const [, setValue] = useState(undefined);
      return (
        <DateTimePicker
          hint="This is a super description"
          error="Very very very very very very very long error"
          label="Error"
          name="datetimepicker"
          onChange={(e) => setValue(e)}
          clearLabel="Clear"
          onClear={() => setValue(undefined)}
        />
      );
    }}
  </Story>
</Canvas>

## Initial data

<Canvas>
  <Story name="initial data">
    {() => {
      const [value, setValue] = useState(new Date('2021-10-13T10:00:00.000Z'));
      return (
        <DateTimePicker
          label="Date time picker"
          name="datetimepicker"
          onChange={(e) => setValue(e)}
          clearLabel="Clear"
          onClear={() => setValue(undefined)}
          value={value}
          hint="This is a super description"
          selectedDateLabel={() => `Date picker, current is undefined`}
        />
      );
    }}
  </Story>
</Canvas>

## Props

<ArgsTable of={DateTimePicker} />
