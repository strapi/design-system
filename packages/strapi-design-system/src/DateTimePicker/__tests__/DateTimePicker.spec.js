import * as React from 'react';
import { render, screen } from '@testing-library/react';
import { DateTimePicker } from '../DateTimePicker';
import { ThemeProvider } from '../../ThemeProvider';
import { lightTheme } from '../../themes';

const setupComponent = (props) => (
  <ThemeProvider theme={lightTheme}>
    <DateTimePicker {...props} />
  </ThemeProvider>
);

const sharedProps = {
  onChange() {},
  name: 'datetimepicker',
  label: 'Date time picker',
  hint: 'This is a super description',
};

const snapshotProps = {
  value: new Date('2021-10-13T10:00:00.000Z'),
  ...sharedProps,
};

const initializationProps = {
  value: new Date('2021-10-13T15:45:00.000Z'),
  step: 15,
  ...sharedProps,
};

const withNewValueInitialProps = {
  value: new Date('2021-10-13T13:43:00.000Z'),
  step: 15,
  ...sharedProps,
};

const withNewValueFinalProps = {
  value: new Date('2021-10-04T13:00:00.000Z'),
  step: 15,
  ...sharedProps,
};

const withEmptyValuesInitialProps = {
  value: new Date('2021-10-13T15:43:00.000Z'),
  step: 15,
  ...sharedProps,
};

const withEmptyValuesFinalProps = {
  step: 15,
  ...sharedProps,
};

describe('DateTimePicker', () => {
  it('snapshots the component', () => {
    const { container } = render(setupComponent(snapshotProps));

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should initialize the two inputs', () => {
    render(setupComponent(initializationProps));

    expect(screen.getByText('15:45')).toBeInTheDocument();
    const datepicker = screen.getByTestId('datetimepicker-date');

    expect(datepicker.value).toBe('10/13/2021');
  });

  it('should rerender a new value passed as props', () => {
    const { rerender } = render(setupComponent(withNewValueInitialProps));

    rerender(setupComponent(withNewValueFinalProps));

    const datepicker = screen.getByTestId('datetimepicker-date');

    expect(datepicker.value).toBe('10/4/2021');
  });

  it('should rerender an empty value if it is passed as props', () => {
    const { rerender } = render(setupComponent(withEmptyValuesInitialProps));

    rerender(setupComponent(withEmptyValuesFinalProps));

    const datepicker = screen.getByTestId('datetimepicker-date');

    expect(datepicker.value).toBe('');
  });
});
