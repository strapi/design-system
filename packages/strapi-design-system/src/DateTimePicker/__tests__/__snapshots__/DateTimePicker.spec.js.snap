// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DateTimePicker snapshots the component 1`] = `
.c7 {
  padding-right: 8px;
  padding-left: 12px;
}

.c14 {
  padding-left: 12px;
}

.c16 {
  padding-right: 16px;
  padding-left: 16px;
}

.c0 {
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c3 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.c5 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c1 > * {
  margin-top: 0;
  margin-bottom: 0;
}

.c1 > * + * {
  margin-top: 4px;
}

.c10 > * {
  margin-top: 0;
  margin-bottom: 0;
}

.c4 > * {
  margin-left: 0;
  margin-right: 0;
}

.c4 > * + * {
  margin-left: 8px;
}

.c2 {
  font-size: 0.75rem;
  line-height: 1.33;
  font-weight: 600;
  color: #32324d;
}

.c17 {
  font-size: 0.875rem;
  line-height: 1.43;
  display: block;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  color: #32324d;
}

.c20 {
  font-size: 0.75rem;
  line-height: 1.33;
  color: #666687;
}

.c9 {
  border: none;
  border-radius: 4px;
  padding-bottom: 0.65625rem;
  padding-left: 0;
  padding-right: 16px;
  padding-top: 0.65625rem;
  color: #32324d;
  font-weight: 400;
  font-size: 0.875rem;
  display: block;
  width: 100%;
  background: inherit;
}

.c9::-webkit-input-placeholder {
  color: #8e8ea9;
  opacity: 1;
}

.c9::-moz-placeholder {
  color: #8e8ea9;
  opacity: 1;
}

.c9:-ms-input-placeholder {
  color: #8e8ea9;
  opacity: 1;
}

.c9::placeholder {
  color: #8e8ea9;
  opacity: 1;
}

.c9[aria-disabled='true'] {
  color: inherit;
}

.c9:focus {
  outline: none;
  box-shadow: none;
}

.c6 {
  border: 1px solid #dcdce4;
  border-radius: 4px;
  background: #ffffff;
  outline: none;
  box-shadow: 0;
  -webkit-transition-property: border-color,box-shadow,fill;
  transition-property: border-color,box-shadow,fill;
  -webkit-transition-duration: 0.2s;
  transition-duration: 0.2s;
}

.c6:focus-within {
  border: 1px solid #4945ff;
  box-shadow: #4945ff 0px 0px 0px 2px;
}

.c8 {
  border: none;
  background: transparent;
  border-radius: 4px;
}

.c8 svg path {
  fill: #8e8ea9;
}

.c12 {
  position: absolute;
  left: 0;
  right: 0;
  bottom: 0;
  top: 0;
  width: 100%;
  background: transparent;
  border: none;
}

.c12:focus {
  outline: none;
}

.c12[aria-disabled='true'] {
  cursor: not-allowed;
}

.c11 {
  position: relative;
  border: 1px solid #dcdce4;
  padding-right: 12px;
  border-radius: 4px;
  background: #ffffff;
  overflow: hidden;
  min-height: 2.5rem;
  outline: none;
  box-shadow: 0;
  -webkit-transition-property: border-color,box-shadow,fill;
  transition-property: border-color,box-shadow,fill;
  -webkit-transition-duration: 0.2s;
  transition-duration: 0.2s;
}

.c11:focus-within {
  border: 1px solid #4945ff;
  box-shadow: #4945ff 0px 0px 0px 2px;
}

.c18 {
  background: transparent;
  border: none;
  position: relative;
  z-index: 1;
}

.c18 svg {
  height: 0.6875rem;
  width: 0.6875rem;
}

.c18 svg path {
  fill: #666687;
}

.c19 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  background: none;
  border: none;
}

.c19 svg {
  width: 0.375rem;
}

.c13 {
  width: 100%;
}

.c15 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c15 svg {
  height: 1rem;
  width: 1rem;
}

.c15 svg path {
  fill: #8e8ea9;
}

<div
  aria-labelledby="datetime-label-1"
  role="group"
>
  <div
    class="c0 c1"
    spacing="1"
  >
    <label
      class="c2"
      for="field-3"
      id="datetime-label-1"
    >
      <div
        class="c3"
      >
        Date time picker
      </div>
    </label>
    <div
      class="c3 c4"
      spacing="2"
    >
      <div
        class=""
      >
        <div>
          <div>
            <div
              class="c0 c1"
              spacing="1"
            >
              <div
                class="c5 c6"
              >
                <div
                  class="c7"
                >
                  <button
                    aria-disabled="false"
                    aria-label="Date picker, current is 10/13/2021"
                    class="c8"
                    type="button"
                  >
                    <svg
                      aria-hidden="true"
                      fill="none"
                      height="1em"
                      viewBox="0 0 24 24"
                      width="1em"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        clip-rule="evenodd"
                        d="M3.869 2.99V0h2.9v2.99h10.463V0h2.9v2.99h.629c2.768 0 3.203.498 3.239 2.926V21c0 2.124-.191 3-2.802 3H2.818C.208 24 0 23.363 0 20.785V6.21c.035-2.049.233-3.22 3.001-3.22h.868zM2.32 20.369c0 .811.245.865.776.865h17.905c.53 0 .68-.012.68-.825V8.233c-.015-.627-.219-.737-.631-.737H2.907c-.413 0-.592.09-.587.573v12.3z"
                        fill="#212134"
                        fill-rule="evenodd"
                      />
                    </svg>
                  </button>
                </div>
                <input
                  aria-autocomplete="none"
                  aria-disabled="false"
                  aria-invalid="false"
                  aria-label="Date time picker"
                  class="c9"
                  data-testid="datetimepicker-date"
                  id="datepicker-4"
                  name="datetimepicker"
                  value="10/13/2021"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
      <div>
        <div
          class="c0 c10"
        >
          <div
            class="c3 c11"
          >
            <button
              aria-disabled="false"
              aria-expanded="false"
              aria-haspopup="listbox"
              aria-label="Date time picker"
              aria-labelledby="timepicker-5-label timepicker-5-content"
              class="c12"
              data-testid="datetimepicker-time"
              id="timepicker-5"
              type="button"
            />
            <div
              class="c5 c13"
            >
              <div
                class="c3"
              >
                <div
                  aria-hidden="true"
                  class="c14"
                >
                  <div
                    class="c15"
                  >
                    <svg
                      fill="none"
                      height="1em"
                      viewBox="0 0 24 24"
                      width="1em"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        clip-rule="evenodd"
                        d="M0 12C0 5.37 5.364 0 11.994 0S24 5.37 24 12s-5.376 12-12.006 12S0 18.63 0 12zm2.4 0c0 5.304 4.296 9.6 9.6 9.6 5.304 0 9.6-4.296 9.6-9.6 0-5.304-4.296-9.6-9.6-9.6A9.597 9.597 0 002.4 12zm8.4-6h1.8v6.3l5.4 3.204-.906 1.476L10.8 13.2V6z"
                        fill="#212134"
                        fill-rule="evenodd"
                      />
                    </svg>
                  </div>
                </div>
                <div
                  class="c16"
                >
                  <span
                    class="c17"
                    id="timepicker-5-content"
                  >
                    10:00
                  </span>
                </div>
              </div>
              <div
                class="c3"
              >
                <button
                  aria-hidden="true"
                  class="c14 c18 c19"
                  tabindex="-1"
                  title="select"
                  type="button"
                >
                  <svg
                    fill="none"
                    height="1em"
                    viewBox="0 0 14 8"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      clip-rule="evenodd"
                      d="M14 .889a.86.86 0 01-.26.625L7.615 7.736A.834.834 0 017 8a.834.834 0 01-.615-.264L.26 1.514A.861.861 0 010 .889c0-.24.087-.45.26-.625A.834.834 0 01.875 0h12.25c.237 0 .442.088.615.264a.86.86 0 01.26.625z"
                      fill="#32324D"
                      fill-rule="evenodd"
                    />
                  </svg>
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <p
      class="c20"
      id="field-3-hint"
    >
      This is a super description
    </p>
  </div>
</div>
`;
